### PR TITLE
Guide Introduction

### PR DESCRIPTION
**Pull request by ajusa**
Created a new `guide` folder that has a guide written out in Markdown. Not much to see there, but I also updated `tester.nim` so that it tries to compile all of the snippets in the guide. Putting this out there for feedback on if this is the "correct" way of doing things.

### CODE REVIEW

**Comment by ajusa** [comment]
Not sure why the other commits I've made for the CI are in this PR, I know I merged everything back from this repository after the PR was accepted, shouldn't that have gotten rid of those commits? If anyone knows how I messed up, please tell me!

**Comment by Araq** [comment]
Nim's documentation generator supports testing of Nim snippets and Markdown.

**Comment by ajusa** [comment]
Yes, but I'm currently writing this up in a flat markdown file with the intent of it eventually being read into a static site generator. The only reason I'd prefer not to use Nim's documentation generator is because it can't compile the Nim snippets into actual runnable js. Willyboar and I were planning on trying to have a static site generator that would also output a running example, similar to other Javascript framework documentation.

**Comment by Araq** [comment]
Hmm I'm pretty sure we do support "nim js" too as the command to use for testing example code.

**Comment by ajusa** [comment]
That's not what I meant, I should have phrased it better sorry. I meant having the guide actually include the generated JS code and running it, as a live example. For example, if I have code that shows how a for loop works in Karax, it would be nice if the actual generated code had the generated output of running that code right below it. 

For example, on the Vue.js guide (https://vuejs.org/v2/guide/#Handling-User-Input), their user input section has a working example, where if you click the button it reverses the string. That way anyone who is just looking through the documentation doesn't need to copy and paste each example to see what it does, the site just includes it.

That's the primary reason I decided against Nim's documentation gen -  it isn't well equipped to handle something like that (as far as I am aware) nor should it be. It's a pretty specialized scenario.

Open to your thoughts on the matter as well! The current setup does what `nim doc` would do, since it does try and compile each code snippet using the `js` backend, and errors if any snippets don't compile. It's also a lot easier to just read and write flat markdown files.

**Comment by Araq** [comment]
> That's the primary reason I decided against Nim's documentation gen - it isn't well equipped to handle something like that (as far as I am aware) nor should it be. It's a pretty specialized scenario.

Ah, yes. Good ideas, please go ahead! :-)

**Comment by ajusa** [comment]
Ready for this to be merged - the guide isn't complete by any means, but this is a good enough start to the point where it should be useful. 

Please let me know if there are any changes I should make! As far as future additions go, I plan to have a section on wrapping external modules, more advanced examples, the virtual dom, and routing.

The guide generation with live examples I talked about will need to be in a future PR, that's out of the scope of this for now. 
As it is, I think that the guide is useful enough to be browsed directly from Github, the guide site generation would just make it better.

File: tests/tester.nim [hunk 1]
```diff
@@ -1,12 +1,29 @@
...
+    var last = 0
+    var trash = ""
+    const startDelim = "```nim"
+    const endDelim = "```"
+    while true:
+      last += contents.parseUntil(trash, startDelim, last)
+      if last == contents.len: break # no matches found
+      var code = ""
+      last += contents.parseUntil(code, endDelim, last+startDelim.len)
+      var res = execCmdEx("nim js -", input = code)
```
Comment (narimiran) [inline tests/tester.nim:22]: In Nim versions 1.2.x and below, the `execCmdEx` proc doesn't have the `input` parameter. This means that Karax is failing on our CIs for v1.2.

Can you work around this limitation?

Comment (ajusa) [reply tests/tester.nim:22]: Sorry about that, I had no idea it failed on 1.2
I'll try and get to it this weekend, should be possible to work around as all I'm doing is checking the output of a command with some stdin, older versions of Nim probably have a more verbose way of accomplishing that.

**Pull request merged**